### DISCUSSION TITLE
docker

### DISCUSSION THREAD

**Discussion by geigerzaehler242** [nlohmann/json#2816] [Q&A] [answered]
hi how can i use this library with docker? below is my dockerfile and also my cmake.txt file i am trying to use:

it all works fine locally for me but i can't figure out how to make this library work on docker.

```dockerfile
#dockerfile
FROM ubuntu:groovy

RUN apt-get update && \
        apt-get install -y build-essential git cmake autoconf libtool pkg-config

RUN git clone https://github.com/nlohmann/json

WORKDIR /submission

COPY CMakeLists.txt main.cpp sample_input.json ./

RUN cmake . && make

CMD ["./sde-test-solution", "sample_input.json", "output_file.json"]
```

```cmake
#cmake.txt

cmake_minimum_required(VERSION 3.11)
project(sde-test-solution)

set(CMAKE_CXX_STANDARD 20)


find_package(nlohmann_json 3.2.0 REQUIRED)
target_link_libraries(sde-test-solution PRIVATE nlohmann_json::nlohmann_json)

add_executable(sde-test-solution main.cpp)
```

**Comment by nlohmann** [comment, marked as answer]
You need to install the library to use it in CMake like this, see https://github.com/nlohmann/json#cmake. This is independent to Docker.

**Reply by geigerzaehler242** [reply]
my make is now:

cmake_minimum_required(VERSION 3.11)
project(sde-test-solution)

set(CMAKE_CXX_STANDARD 20)

add_executable(sde-test-solution main.cpp Bond.cpp)

include(FetchContent)

FetchContent_Declare(json
  GIT_REPOSITORY https://github.com/nlohmann/json.git
  GIT_TAG v3.7.3)

FetchContent_GetProperties(json)
if(NOT json_POPULATED)
  FetchContent_Populate(json)
  add_subdirectory(${json_SOURCE_DIR} ${json_BINARY_DIR} EXCLUDE_FROM_ALL)
endif()

target_link_libraries(sde-test-solution PRIVATE nlohmann_json::nlohmann_json)


im getting the following error though:

#9 158.0 [ 33%] Building CXX object CMakeFiles/sde-test-solution.dir/main.cpp.o
#9 159.5 In file included from /submission/main.cpp:12:
#9 159.5 /submission/_deps/json-src/single_include/nlohmann/json.hpp: In static member function 'static CharType nlohmann::detail::binary_writer<BasicJsonType, CharType>::to_char_type(uint8_t)':
#9 159.5 /submission/_deps/json-src/single_include/nlohmann/json.hpp:12595:28: warning: 'template<class _Tp> struct std::is_pod' is deprecated: use is_standard_layout && is_trivial instead [-Wdeprecated-declarations]
#9 159.5 12595 |         static_assert(std::is_pod<CharType>::value, "CharType must be POD");
#9 159.5       |                            ^~~~~~
#9 159.5 In file included from /usr/include/c++/10/bits/move.h:57,
#9 159.5                  from /usr/include/c++/10/bits/nested_exception.h:40,
#9 159.5                  from /usr/include/c++/10/exception:148,
#9 159.5                  from /usr/include/c++/10/ios:39,
#9 159.5                  from /usr/include/c++/10/istream:38,
#9 159.5                  from /usr/include/c++/10/fstream:38,
#9 159.5                  from /submission/main.cpp:9:
#9 159.5 /usr/include/c++/10/type_traits:697:5: note: declared here
#9 159.5   697 |     is_pod
#9 159.5       |     ^~~~~~
#9 159.9 /submission/main.cpp: In function 'int main(int, const char**)':
#9 159.9 /submission/main.cpp:257:15: error: 'setw' is not a member of 'std'
#9 159.9   257 |     o << std::setw(4) << outputJson << std::endl;
#9 159.9       |               ^~~~
#9 161.1 make[2]: *** [CMakeFiles/sde-test-solution.dir/build.make:63: CMakeFiles/sde-test-solution.dir/main.cpp.o] Error 1

**Reply by nlohmann** [reply]
Can you please try the latest release version, 3.9.1.

**Reply by nlohmann** [reply]
Without knowing the contents of `main.cpp`, I cannot say more.

**Reply by geigerzaehler242** [reply]
i changed my cmake to specify the 3.9.1 version. 

```cmake
FetchContent_Declare(json
  GIT_REPOSITORY https://github.com/nlohmann/json.git
  GIT_TAG v3.9.1)
```

however i still get this error:

```
#8 39.34 -- Using the single-header code from /submission/_deps/json-src/single_include/
#8 39.36 -- Found Git: /usr/bin/git (found version "2.27.0") 
#8 39.36 -- Operating system: Linux-5.10.25-linuxkit; Linux buildkitsandbox 5.10.25-linuxkit #1 SMP Tue Mar 23 09:27:39 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
#8 39.38 -- Compiler: c++ (Ubuntu 10.2.0-13ubuntu1) 10.2.0; Copyright (C) 2020 Free Software Foundation, Inc.; This is free software; see the source for copying conditions.  There is NO; warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#8 39.39 -- Configuring done
#8 39.55 -- Generating done
#8 39.56 -- Build files have been written to: /submission
#8 39.60 make[1]: Entering directory '/submission'
#8 39.61 make[2]: Entering directory '/submission'
#8 39.63 Scanning dependencies of target sde-test-solution
#8 39.63 make[2]: Leaving directory '/submission'
#8 39.63 make[2]: Entering directory '/submission'
#8 39.64 [ 50%] Building CXX object CMakeFiles/sde-test-solution.dir/main.cpp.o
#8 41.96 /submission/main.cpp: In function 'int main(int, const char**)':
#8 41.96 /submission/main.cpp:257:15: error: 'setw' is not a member of 'std'
#8 41.96   257 |     o << std::setw(4) << outputJson << std::endl;
#8 41.96       |               ^~~~
#8 43.61 make[2]: Leaving directory '/submission'
#8 43.61 make[2]: *** [CMakeFiles/sde-test-solution.dir/build.make:63: CMakeFiles/sde-test-solution.dir/main.cpp.o] Error 1
#8 43.61 make[1]: Leaving directory '/submission'
#8 43.61 make[1]: *** [CMakeFiles/Makefile2:345: CMakeFiles/sde-test-solution.dir/all] Error 2
#8 43.61 make: *** [Makefile:130: all] Error 2
------
executor failed running [/bin/sh -c cmake . && make]: exit code: 2
```

**Reply by geigerzaehler242** [reply]
i don't know where this "o << std::setw(4) << outputJson << std::endl;" is coming from. it must be in your libraries main because it is not in my main

**Reply by geigerzaehler242** [reply]
apparently i had to add the following include for some reason in my main:

```cpp
#include <iomanip>
```

**Reply by nlohmann** [reply]
Do you need further assistance with this issue?
